content/posts/0006-java-rearchitecture.md

```diff
@@ -0,0 +1,194 @@
+
+`pack build samples/java --path java/jar`
+
+You can then enable a feature at run time using the `BPL_<feature-name>_ENABLED` environment variable detailed for each feature. Further variables can then be used to customise the default configuration for this feature.
+
+#### **Java Native Memory Tracking (NMT)**  
+
+Native Memory Tracking (NMT) is a feature that tracks internal memory usage in the JVM. Support for this feature has been added to the Paketo Java Buildpack and is now enabled by default. When the JVM exits, it will print a summary of memory usage. In addition, you can use the JDK tool ‘jcmd’ to dump and view the data collected (see the [JVM Type configuration][install-jvm-type] section for details on how to ensure you have a JDK at runtime). There are two runtime configuration options for this feature:
+
+`BPL_JAVA_NMT_ENABLED` - set to `true` by default, however the tracking can cause a 5-10% performance overhead, so this variable can be set to `false` to disable memory tracking when the JVM starts.
+
+`BPL_JAVA_NMT_LEVEL` - set to `summary` by default, this specifies the level of detail for the memory usage data captured. For a more detailed output, it can be set to `detail`.
+
+The following default NMT arguments are passed to the JVM:
+
+```
+-XX:+UnlockDiagnosticVMOptions
+-XX:NativeMemoryTracking=summary
+-XX:+PrintNMTStatistics`
+```
+
+Example:
+
+To start a new bash session and capture NMT data from a running container run the following:
+
+`docker exec -it <container-id> /cnb/lifecycle/launcher /bin/bash`
+
+If a JDK was installed, this session should have the JDK tool `jcmd` on the $PATH variable. To capture NMT data you can run:
+
+`jcmd 1 VM.native_memory summary`
+
+NMT data will also be printed at JVM exit:
+
+![NMT](/images/posts/0006/nmt.png)
+
+#### **Java Flight Recorder (JFR)**
+
+Java Flight Recorder is a tool which is integrated into the JVM and allows for the collection of profiling and diagnostic data about your application. Support for this feature has been added, however, it is disabled by default. To enable it, simply use the environment variable:
+
+`BPL_JFR_ENABLED` - set to `false` by default
+ 
+The tool supports a range of configuration options which can be set using the following environment variable:
+
+`BPL_JFR_ARGS` - this supports a comma-separated list of options (full list detailed in the Oracle docs [here](https://docs.oracle.com/javacomponents/jmc-5-4/jfr-runtime-guide/comline.htm#JFRUH193)). If the tool is enabled but no options are specified here, the buildpack will at a minimum configure the tool to write data to a temporary file on JVM exit. The following default options will be configured to allow this:
+
+`dumponexit=true` - this tells the JFR tool to write recording data to a file when the JVM exits.
```

Comment:
Something funky is happening with the format of this stuff:

![image](https://user-images.githubusercontent.com/21328286/146451132-88608fb7-9dda-4280-8112-cdf00302ad88.png)

Maybe some words or formatting is missing?

Reply:
Thanks for the review! I don't think there's anything missing here but which part seems funky? If it's the args, I could add bullet points before the two, to indicate they belong to BPL_JFR_ARGS?

Reply:
Oh, yeah, that clarifies things. I was seeing all three as siblings, but you're saying `dumponexit` and `filename=` are children of `BPL_JFR_ARGS`? Some bulleting would help make that more clear to me 🙂

Reply:
Nice catch, I was looking at it for too long so i didn't notice this 😬